feat: add cleanup-branches action (#1334)
grafana/shared-workflows: actions/cleanup-branches/README.md

# cleanup-branches

> Composite action (step) to get create github app token using vault. This action will query for for branches that are not in an open PR, and will delete them if 'dry-run' is 'false'. Protected branches are excluded as well.

Reviewer: ```suggestion
Composite action (step) to get create GitHub app token using Vault. This action will query for for branches that are not in an open PR, and will delete them if `dry-run` is `false`. Protected branches are excluded as well.
```

## Inputs

| Name             | Type   | Description                                                                                                                                                                    | Default Value         | Required |
| ---------------- | ------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | --------------------- | -------- |

> | `token`          | string | GitHub token used to authenticate with `gh`. Requires permission to query for protected branches and delete branches (contents: write) and pull requests (pull_requests: read) | `${{ github.token }}` | true     |

Reviewer: ```suggestion
| `token`          | `string` | GitHub token used to authenticate with `gh`. Requires permission to query for protected branches and delete branches (`contents: write`) and pull requests (`pull_requests: read`) | `${{ github.token }}` | `true`     |
```

> | `dry-run`        | bool   | If 'dry-run' is not 'true', then the action will print branches to be deleted, but will not delete them                                                                        | `"true"`              | true     |

Reviewer 2: ```suggestion
| `dry-run`        | bool   | If 'dry-run' is 'true', then the action will print branches to be deleted, but will not delete them                                                                        | `"true"`              | true     |
```

> | `dry-run`        | bool   | If 'dry-run' is not 'true', then the action will print branches to be deleted, but will not delete them                                                                        | `"true"`              | true     |

Reviewer: ```suggestion
| `dry-run`        | `bool`   | If 'dry-run' is `'true'`, then the action will print branches to be deleted, but will not delete them                                                                        | `'true'`              | `true`     |
```

Author: composite actions don't really support bools; should it still be listed as a string in this table?

## Examples

### Clean up branches on a weekly cron schedule

<!-- x-release-please-start-version -->

```yaml
name: Clean up orphaned branches
on:
  schedule:
    - cron: '0 9 * * 1'

jobs:

>   main:

Reviewer: ```suggestion
  cleanup-branches:
```

# cleanup-branches

> Composite action (step) to query for for branches that are not in an open PR, and delete them if 'dry-run' is 'false'. Protected branches are excluded as well.

Reviewer 2: ```suggestion
Composite action (step) to query for branches that are not in an open PR, and delete them if 'dry-run' is 'false'. Protected branches are excluded as well.
```

| `token`          | `string` | GitHub token used to authenticate with `gh`. Requires permission to query for protected branches and delete branches (`contents: write`) and pull requests (`pull_requests: read`)   | `${{ github.token }}` | true     |
| `dry-run`        | `bool`   | If `'true'`, then the action will print branches to be deleted, but will not delete them                                                                                             | `'true'`              | true     |
| `max-date`       | `string` | Value passed to `date -d`; a human readable date string. Maximum date of the head ref of a branch in order to be deleted.                                                            | `"2 weeks ago"`       | false    |

## Examples

### Clean up branches on a weekly cron schedule

<!-- x-release-please-start-version -->

```yaml
name: Clean up orphaned branches
on:
  schedule:
    - cron: '0 9 * * 1'

jobs:
  cleanup-branches::
    runs-on: ubuntu-latest
    permissions:
      contents: write
      pull-requests: read
    steps:

>       - uses: actions/checkout@v5

Reviewer 2: i'm not sure this will fetch all the branches that should be evaluated/deleted by this action. i believe it only fetches the triggering ref by default, so when you do `git branch -r`, it is not seeing the full set of branches this action is intended to evaluate and delete
